### PR TITLE
refactor(smartengine): add abstraction layer for WASM engine

### PR DESCRIPTION
**Pull request by xxchan**
This PR contains:
- refactor to add an abstraction layer of wasm engine. And refactor wasmtime engine to use the common code. 
- add WasmEdge engine

To run WasmEdge test, install wasmedge: https://wasmedge.org/book/en/quick_start/install.html
``
cargo test -p fluvio-smartengine --features wasmedge-engine
``

I'd like to gain feedbacks about the overall design first. If the general ideas look good, we can take a deeper look and begin to resolve other points (e.g., styling, CI, ...), and maybe split into smaller PRs if necessary.

Some notes:
- Most code & tests are shared by both engines. Only folder `wasmedge` is WasmEdge-specific code, and that's ~300 lines of simple code. To discuss how to maintain WasmEdge related code, and to get the ideas of how the common code work, it might be better to keep WasmEdge related code now. This can actually make it easier to review this PR. I can remove the WasmEdge related code after the code review if that's not decided.
- Since there are some re-org. Maybe it's easier to review by checking out the branch and review locally. (Just a suggestion)

### CODE REVIEW

**Comment by sehz** [comment]
Nice work on adding WasmEdge features.  To make it easier to review, can you split these into two separate PR: (1) non WasEdge related changes so we can merge that (2) WasEdge related code.   We still have to figure out how to support WasEdge features with minimum  impact to Fluvio developers

**Comment by xxchan** [comment]
WasmEdge code is removed.

**Comment by xxchan** [comment]
> can you reorder terms in of:
> 
> * std
> * third party
> * fluvio dep

https://github.com/xxchan/fluvio/pull/1#discussion_r1191642981

@sehz Just FYI, there's a rustfmt option (unstable) https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#group_imports

I added
```
group_imports = "StdExternalCrate"
```
and run
```
cargo +nightly fmt -p fluvio-smartengine 
```

Not sure whether you want it for the whole repo, as you use stable toolchian.

File: crates/fluvio-smartengine/src/engine/wasmtime/instance.rs [hunk 1]
```diff
@@ -1,90 +1,93 @@
...
+    instance: Instance,
+    records_cb: Arc<RecordsCallBack>,
+    params: SmartModuleExtraParams,
+    version: i16,
+}
+pub struct WasmtimeContext {
+    pub state: WasmState,
 }
 
-impl SmartModuleInstance {
```
Comment (xxchan) [inline crates/fluvio-smartengine/src/engine/wasmtime/instance.rs:26]: This goes to `common/mod.rs`

File: crates/fluvio-smartengine/src/engine/wasmtime/instance.rs [hunk 2]
```diff
@@ -118,75 +121,17 @@ impl SmartModuleInstanceContext {
...
-            .get()
-            .and_then(|m| m.copy_memory_from(store).ok())
-            .unwrap_or_default();
-        let mut output = D::default();
-        output.decode(&mut std::io::Cursor::new(bytes), self.version)?;
-        Ok(output)
-    }
-}
-
-pub(crate) trait SmartModuleTransform: Send + Sync {
```
Comment (xxchan) [inline crates/fluvio-smartengine/src/engine/wasmtime/instance.rs:157]: This goes to `common/transforms/mod.rs`

File: crates/fluvio-smartengine/Cargo.toml [hunk 3]
```diff
@@ -17,12 +17,12 @@ wasi = ["wasmtime-wasi", "engine"]
...
 default = ["engine"]
 
-
 [dependencies]
 tracing ={ workspace = true }
 thiserror = { workspace = true }
 anyhow = { workspace = true }
-serde = { workspace = true,  features = ['derive'] }
+# std is needed for ser/de AtomicU64
+serde = { workspace = true,  features = ['derive', 'std'] }
```
Comment (xxchan) [inline crates/fluvio-smartengine/Cargo.toml:25]: previously `std` features is depended by `wasmtime`, I added it when developing WasmEdge

File: crates/fluvio-smartengine/src/engine/wasmtime/engine.rs [hunk 4]
```diff
@@ -31,71 +28,38 @@ impl SmartEngine {
...
 
-impl Debug for SmartEngine {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "SmartModuleEngine")
-    }
-}
-
-/// Building SmartModule
-#[derive(Default)]
-pub struct SmartModuleChainBuilder {
```
Comment (xxchan) [inline crates/fluvio-smartengine/src/engine/wasmtime/engine.rs:42]: This becomes `src/engine/mod.rs`

File: crates/fluvio-smartengine/src/engine/common/init.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        &mut self,
+        input: SmartModuleInitInput,
+        instance: &mut I,
+        ctx: &mut I::Context,
+    ) -> Result<()>
+    where
+        I: WasmInstance,
+        F: WasmFn<Context = I::Context>,
+    {
+        let (ptr, len, version) = instance.write_input(&input, ctx)?;
```
Comment (sehz) [inline crates/fluvio-smartengine/src/engine/common/init.rs:46]: the prev version used slice abstraction.  is there a reason for changing this?

File: crates/fluvio-smartengine/src/engine/common/mod.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    fn params(&self) -> SmartModuleExtraParams;
+
+    fn get_fn(&self, name: &str, ctx: &mut Self::Context) -> Result<Option<Self::Func>>;
+
+    fn write_input<E: Encoder>(
+        &mut self,
+        input: &E,
+        ctx: &mut Self::Context,
+    ) -> Result<(i32, i32, i32)>;
```
Comment (sehz) [inline crates/fluvio-smartengine/src/engine/common/mod.rs:23]: Use slice abstraction as before

**Review by sehz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**